### PR TITLE
Async functions

### PR DESCRIPTION
**Pull request by MartinDawson**
### Context
Async functions are needed where data does not resolve immediately, e.g. `=PRICE("AMZN")`

### How has this been tested?
- Refactored all tests to pass with the new async syntax
- Custom-functions.spec.ts 2 new tests for timeout & generic errors were added.
- Other existing tests cover other async cases

### Types of changes
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Additional language file or change to the existing one (translations)

Added `!TIMEOUT#` error. Needs translations still

### Related issue(s):
1. https://github.com/handsontable/hyperformula/issues/9

### Checklist:
- [ ] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I described the modification in the CHANGELOG.md file.

- Code needs formatting with prettier or something automatically.
- Public function tsdoc comments need updating in `Hyperformula.ts`. Not sure how you guys want this done, i.e this is the old way: `   * const hfInstance = HyperFormula.buildFromArray(sheetData, { maxColumns: 1000 });`
However if we just do this: `* const hfInstance = await HyperFormula.buildFromArray(sheetData, { maxColumns: 1000 });` it isn't going to compile by default. Is that okay? Or should we wrap this in a self calling async function for the tsdoc comments?

Outstanding issues on this PR:

Please see here for full context: https://github.com/handsontable/hyperformula/issues/9

- Parallel evaluation (not sure how too do this, and it seems like an enhancement anyway that can be a seperate PR).
- Circuit breaker - Need clarification if you guys want this functionality.
- Retries - Need clarification if you guys want this functionality. We can perhaps do this behind the scenes if there's a TIMEOUT error?
- Browser tests are not passing. Reason is that jasmine does not have the jest specific `.resolves` and `.rejects` property that jest needs. What do you think the best way of handling this is? I had a look at the code to see if we can add it to jasmine bootstrap file but there is a lot of code jest uses internally to unwrap the promises.

### CODE REVIEW

**Review by wojciechczerniak — COMMENTED** [review]
Thank you! Looks great. Is it current implementation you're using for your project? How well does it work in a real use case?

My notes:
- It's a huge breaking change, HyperFormula v2.0 
- I wouldn't worry about translations until this PR is close to completion
- Lots of documentation will have to be adjusted to async methods
- All examples have to be adjusted with new public methods (as noted in PR desc)
- Custom functions documentation has to be rewritten
- I won't be able to review it completely, calculation engine will require @bardek8 @izulin or @voodoo11 expertise

> Browser tests are not passing. Reason is that jasmine does not have the jest specific .resolves and .rejects property that jest needs. What do you think the best way of handling this is? I had a look at the code to see if we can add it to jasmine bootstrap file but there is a lot of code jest uses internally to unwrap the promises.

Those are Jest specific expectations/matches? Maybe we can map them for Jasmine as we did for some other matchers: https://github.com/handsontable/hyperformula/blob/master/test/_setupFiles/matchers/toMatchObject.ts

> Code needs formatting with prettier or something automatically.

HF does not have such a tool yet, sorry. EditorConfig with IDE formatting works well though.

**Comment by wojciechczerniak** [comment]
@MartinDawson pls change target branch to `develop`

File: test/column-range.spec.ts [hunk 1]
```diff
@@ -58,23 +58,23 @@ describe('Column ranges', () => {
     expect(engine.graph.existsEdge(h42, dg)).toBe(false)
   })
 
-  it('should clear column range set in graph when removing column', () => {
-    const engine = HyperFormula.buildFromArray([
+  it('should clear column range set in graph when removing column', async() => {
+const engine = await HyperFormula.buildFromArray([
```
[the remark this answers is not in the document]
Comment (MartinDawson) [reply test/column-range.spec.ts:62]: I had a look at trying to auto fix all files formatting issues. Prettier formatting wasn't working when I tried, maybe it's cause the project has no prettier spec? Not sure but I'll try again

File: src/interpreter/InterpreterValue.ts [hunk 2]
```diff
@@ -10,7 +10,10 @@ export const EmptyValue = Symbol('Empty value')
 export type EmptyValueType = typeof EmptyValue
 export type InternalNoErrorScalarValue = RichNumber | RawNoErrorScalarValue
 export type InternalScalarValue = RichNumber | RawScalarValue
+export type AsyncInternalScalarValue = Promise<InternalScalarValue>
+
 export type InterpreterValue = RichNumber | RawInterpreterValue
+export type AsyncInterpreterValue = Promise<InterpreterValue>
```
[the remark this answers is not in the document]
Comment (MartinDawson) [reply src/interpreter/InterpreterValue.ts:16]: I think so. `InterpreterValue` is used in many places still so we need this.

`AsyncInterpreterValue` was just easier to put here than do `Promise<InterpreterValue>` in 100 different places

**Comment by MartinDawson** [comment]
> Thank you! Looks great. Is it current implementation you're using for your project? How well does it work in a real use case?

Thanks. Yes we need in our project now.

I have released this as a `2.0.0` on my project and am testing it now in a real prod application. I've already noticed some issues I think and will fully test and come back to this PR to log them and see improvements.

**Comment by MartinDawson** [comment]
@wojciechczerniak This pull request is wrong because in this implementation it is waiting until async functions have finished before rendering the cells.

When it should immediatelly render the cells with something like `Loading...` (like google sheets), save the AST & state of the async cells and then do a second run through for the async cells.

I found a way to do this much nicer so this PR I am closing and will create a new PR soon with much smaller breaking changes.

**Pull request closed without merging**